transform

ddudt opened the pull request:
Attempting to merge changes from `transform` into `master`.  It passed all the tests on `dev`.

ddudt committed:
Initial transform commit

Renamed zernike.py to transfrom.py to house the FourierZernikeBasis and Transform classes.
Renamed nodes.py to grid.py to house the CollocationGrid class.

ddudt committed:
Updated filenames in test imports

evanyerger committed:
Merge branch 'io' into dev

ddudt committed:
Grid class and children

Created Grid class structure.
Added LinearGrid and ConcentricGrid child classes.
Updated existing tests.
Updated configuration documentation.

ddudt committed:
initial Basis and Transform structure

Created basis.py to contain Basis classes, and moved relevant code from transform.py (previously zernike.py).
Basis is an abstract parent class, the only child is probably FourierZernikeBasis.
Need to finish separating grid and basis functionality from transform code.

ddudt committed:
more Basis restructuring

evanyerger committed:
Merge branch 'io' into dev

ddudt committed:
cherry-picking from master to updated travis.yml

ddudt committed:
mostly finished with Grid and Basis

Implemented abstract base classes (ABC) for Grid and Basis.
PowerSeries, DoubleFourierSeries, and FourierZernikeBasis are children of Basis.
Added functionality to Basis classes.
Moved some functionality back to transform.py.

ddudt committed:
prepping Transform class

Updated documentation to Grid, Basis, and Transform.
Removed methods related to FFT from Transform!
We will have to re-add FFT functionality later.

dpanici committed:
Obj fxns (#4)

* made ABC for ObjectiveFunction, and subclassed for ForceErrorNodes

* Added AccelErrorNodes Class

* Added basic integration test to check if desc will run successfully on DSHAPE

Co-authored-by: Daniel Dudt <33005725+ddudt@users.noreply.github.com>

dpanici committed:
fixed some input variable names in __main__, still having trouble running DSHAPE

evanyerger committed:
Merge branch 'io' into dev

evanyerger committed:
Fixed merge conflict in __main__.py calling solve_eq_continuation

evanyerger committed:
Added missing lines to InputReader.parse_inputs. inputs[bZ] now loads correctly.

dpanici committed:
Changed integration test from DSHAPE to SOLOVEV

ddudt committed:
Transform methods

Refactored Transform initializer and transformation functions.
Bug fixes in Grid and Basis classes.
Improved documentation.

ddudt committed:
Transform basic functionality is stable

Added testing of Transform functions.
Moved some functionality from backend.py to basis.py.

ddudt committed:
changed nodes to shape (N,3)

Grid.nodes and Grid.volumes used to be shape (3,N).
They are now of shape (N,3) to follow the same format as Basis.nodes.
Also changed node and mode sorting orders: sorts by toroidal first, then poloidal, then radial.

ddudt committed:
naive methods for changing transform resolution

Provided a way to update Transform resolution through setter methods.
Will need to make these more efficient by avoiding recomputing existing matrices.

ddudt committed:
Transform setter method improvements

Fixed node index order bug in Basis class.
== operator overloading.
More efficient Transform setter methods (still room for improvement when changing Basis).
More extensive testing.

dpanici committed:
cherry-picked obj_fxns updates

Cherry-picked from master after pull request #3.
Initial merges of obj_fxns and transform updates.
Grid base class is no longer abstract, LinearGrid resolution notation changed, and added Grid.axis property.
Initial changes to introduce new transform notation into Configuration and ObjectiveFunction uses.

dpanici committed:
Obj fxns (#6)

* made ABC for ObjectiveFunction, and subclassed for ForceErrorNodes

* Added AccelErrorNodes Class

* Added basic integration test to check if desc will run successfully on DSHAPE

* Added basic tests of logic of ObjectiveFunctionFactory

* Forgot to add the test file for SOLOVEV

Co-authored-by: Daniel Dudt <33005725+ddudt@users.noreply.github.com>

evanyerger committed:
Merge branch 'io' into dev

dpanici committed:
Merge branch 'io' into dev

Fixed small name errors that were preventing integration test from running.

ddudt committed:
obj_fxns & transform merge is stable

Major restructuring of the following to accomodate new transform structure: configuration compute function, boundary conditions, objective functions, and continuation script.
Added num_nodes and num_modes properties to to Grid, Basis, Transform classes, along with other minor changes.
Plotting is broken but VMEC error should be intact.
Added second continuation step to SOLOVEV test with toroidal modes.

ddudt committed:
boundary condition bug fix

Always have to use a non-symmetric Grid to compute the BC properly.  This could be changed after symmetry gets implemented in Basis.
Updated initial guess documentation.

ddudt committed:
bug fixes for continuation steps

Fixed bug when changing Transform.grid or Transform.basis.  Reverted to naive method, which surprisingly seems faster (maybe redundant but vectorized is faster than adding logic).
Fixed bug in configuration change_resolution function for when basis changes.
Updated is_nested function to work with new transform structure.

ddudt committed:
replaced numpy with JAX in basis evaluations

Attempting to make JAX run.

Commit:
JAX now pre-computes, but still does not compile

ddudt committed:
jacobi jax bug fix

Accidentally deleted a line in jacobi function when sorting out JAX issues.

ddudt committed:
trying to make BC's work with JAX

Changed Basis function decorators.
Used jax.numpy in more places.
Rewrote boundary condition to not explicity initialize objects.

ddudt committed:
JAX should work now

Bug fix for jax numpy in BC function.
Continuation changes that I thought I already committed.
Hacky plotting comparison to VMEC, but doesn't save data.

ddudt committed:
boundary input format fix for jax

Fixed how spectral boundary inputs are formatted to work reliably with JAX numpy arrays.
Also added jnp to some Grid functions (probably not necessary, but doesn't hurt).

ddudt committed:
removed jnp from Grid methods

It turns out using jnp in the Grid methods does hurt...

ddudt committed:
added VMEC comparison plot

Now plots comparison to initial guess and VMEC solution (if given).
Also made updating the Transform resolution more efficient.

ddudt committed:
added solution error quantification (relative to VMEC)

ddudt committed:
git says I have changes

ddudt committed:
Revert "git says I have changes"

This reverts commit fcefbf38952d5031ef5640a0d6e6e55e47cc5634.

ddudt committed:
resolved initial merge conflicts with io

ddudt committed:
hacky fixes to make it pass tests

Removed perturbation tests since they are outdated.  Will have to write new tests when we create the Perturber class.
Stopped Checkpoint from saving solutions, since there are problems with saving Transform objects.

ddudt committed:
Merge 'transform' into 'dev', with preference to 'transform' in conflicts

ddudt committed:
Merge branch 'dev' into 'transform', strategy=transform

ddudt committed:
reducing jax requirement to <= 0.2.1

JAX 0.2.1 is the latest version that is known to work with our usage.

ddudt committed:
updating equals (==) function with jnp

ddudt committed:
increased absolute tolerance for numpy array testing

Default was atol=0 for np.testing_allclose, increased it to atol=1e-8.

ddudt committed:
further reducing jax requirements

jax <= 0.1.77
jaxlib <= 0.1.55

ddudt committed:
Merge branch 'master' into transform

ddudt committed:
resolving weird merge stuff from master

ddudt commented:
The initial pull request failed some tests because it kept duplicate copies of some functions during the merge.  I manually resolved the issues in the latest commit to `transform`, and it passes now.

dpanici committed:
fixed arguments for compute_lambda_err for non-stellarator symmetric case

evanyerger commented on desc/configuration.py:
```diff
@@ -1,17 +1,143 @@
 
     # TODO: replace zern_idx & lambda_idx with Transform objects
-    def __init__(self, inputs=None, load_from=None, file_format='hdf5'): #bdry, cP, cI, Psi, NFP, zern_idx, lambda_idx, sym=False, x=None, axis=None) -> None:
+    def __init__(self, inputs:dict=None, load_from:str=None, file_format:str='hdf5') -> None:
```
load_from shouldn't be type hinted here: it's supposed to be either a file name OR file instance, so that saving from EquilibriaFamily and Equilibrium can save these without having to open and close the file all the time. I also do type checking in the class, so none is needed at a high-level.

evanyerger commented on desc/configuration.py:
```diff
@@ -72,40 +201,33 @@ def _init_from_inputs_(self, inputs=None):
-        return None
 
-    def _init_from_file_(self, load_from=None, file_format=None):
+    def _init_from_file_(self, load_from:str=None, file_format:str=None) -> None:
```
Remove type hinting. See line 140.

evanyerger commented on desc/configuration.py:
```diff
@@ -208,39 +326,24 @@ def save(self, save_to, file_format='hdf5', file_mode='w'):
+    """
 
-    def __init__(self, inputs=None, load_from=None, file_format='hdf5'):#bdry, cP, cI, Psi, NFP, zern_idx, lambda_idx, sym=False, x=None, axis=None, objective=None, optimizer=None) -> None:
+    def __init__(self, inputs:dict=None, load_from:str=None, file_format:str='hdf5') -> None:
```
Remove type hint on load_from

evanyerger commented on desc/configuration.py:
```diff
@@ -208,39 +326,24 @@ def save(self, save_to, file_format='hdf5', file_mode='w'):
-        return None
 
-    def _init_from_file_(self, load_from=None, file_format=None):
+    def _init_from_file_(self, load_from:str=None, file_format:str=None) -> None:
```
Remove type hint on load_from

evanyerger reviewed (changes requested):
Type hinting load_from might cause tests to fail if it is checked. It's supposed to be dynamically typed anyway, and is checked in the readers and writers. Everything else looks good.

ddudt committed:
updates related to Configuration class

Grid, Basis, Transform: Now using _method_name_() convention for private-ish methods. Added self._save_attrs_ attribute for saving.
Edited Configuration class to have relevant getter & setter fields for Transforms.
Updates to Checkpoint saving and associated plotting routines.

dpanici commented:
Having some issues getting DSHAPE to converge to VMEC solution, though that may just be me not knowing which resolutions to increase ( I mainly just increased Mnodes). Also, when running desc with `-p --vmec` option, the `vmec_error` function that gets called returns an error from transform.py "ValueError: Coefficients dimension is incompatible with transform basis"

evanyerger reviewed (approved):
Requested changes have been resolved.

ddudt committed:
bug fix for vmec comparison

VMEC comparison plot assumed ANSI indexing, fixed to used the saved Basis.
Increased resolution of DSHAPE input file so it converges to VMEC solution.

ddudt commented:
> Having some issues getting DSHAPE to converge to VMEC solution, though that may just be me not knowing which resolutions to increase ( I mainly just increased Mnodes). Also, when running desc with `-p --vmec` option, the `vmec_error` function that gets called returns an error from transform.py "ValueError: Coefficients dimension is incompatible with transform basis"

I resolved both of these in the latest commit.  Increased resolution makes the DSHAPE converge to the correct solution.  The error in the VMEC comparison plot was caused by assuming ANSI indexing -- it can now handle any Basis.

ddudt committed:
bug fix in transform setting grid and basis
